File: src/protocols/ructf_tcp.rs
use super::ProtocolHandler;
use crate::models::flag::{Flag, FlagStatus};
use crate::settings::ProtocolConfig;
use log::info;
use std::collections::HashMap;
use std::ffi::CStr;
use std::io::{Read, Write};
use std::net::TcpStream;

pub struct RuCtfTcp;

impl ProtocolHandler for RuCtfTcp {
    fn send_flags(&self, queue_flags: Vec<Flag>, config: &ProtocolConfig) -> Vec<Flag> {
        let responses: HashMap<String, Vec<&str>> = HashMap::from([
            (
                FlagStatus::QUEUED.to_string(),
                vec![
                    "timeout",
                    "game not started",
                    "try again later",
                    "game over",
                    "is not up",
                    "no such flag",
                ],
            ),
            (
                FlagStatus::ACCEPTED.to_string(),
                vec!["accepted", "congrat"],
            ),
            (
                FlagStatus::REJECTED.to_string(),
                vec![
                    "bad",
                    "wrong",
                    "expired",
                    "unknown",
                    "your own",
                    "too old",
                    "not in database",
                    "already submitted",
                    "invalid flag",
                ],
            ),
        ]);
        match TcpStream::connect(format!("{}:{}", config.checksys_host, config.checksys_port)) {
            Ok(mut stream) => {
                let mut buf: [u8; 4096] = [0; 4096];
                match stream.read(&mut buf) {
                    Ok(_) => {
                        let welcome_msg = std::str::from_utf8(&buf).unwrap().to_lowercase();
                        if !welcome_msg.contains("enter your flags") {
                            return Vec::new();
                        }
                        let mut updated_flags: Vec<Flag> = Vec::new();
                        for mut flag in queue_flags {
                            stream.write((flag.flag).as_bytes()).unwrap();
                            buf.fill(0);
                            match stream.read(&mut buf) {
                                Ok(_) => {
                                    let answ =
                                        CStr::from_bytes_until_nul(&buf).unwrap().to_string_lossy();
                                    let msg =
                                        answ.replace(format!("[{}] ", flag.flag).as_str(), "");
                                    flag.checksystem_response = Some(msg.clone().into());
                                    for (status, key_words) in &responses {
                                        if key_words
                                            .iter()
                                            .any(|word| msg.to_lowercase().contains(word))
                                        {
                                            flag.status = status.clone().into();
                                            break;
                                        }
                                    }
                                    updated_flags.push(flag);
                                }
                                Err(e) => error!("Failed to receive data from check system: {}", e),
                            }
                        }
                        info!("Checksys response: {:?}", updated_flags);
                        return updated_flags;
                    }
                    Err(e) => error!("Failed to receive data from check system: {}", e),
                }
            }
            Err(e) => error!("Failed to connect to check system: {}", e),
        }
        Vec::new()
    }
}

File: flag_senders/protocols/ructf_tcp.py
import socket
import logging

from flag_sender import FlagSender
from farm import Flag, FlagStatus
from typing import List


class RuCtfTcpFlagSender(FlagSender):
    def connect_to_checksystem(self, host: str, port: int) -> socket.socket:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(5.0)
        sock.connect((host, port))

        data = sock.recv(4096)
        if not data:
            raise Exception("Нет данных от сервера журейки.")
        if "Enter your flags" not in data.decode("utf-8"):
            raise Exception(f"Неожиданный ответ: {data.decode('utf-8')}")

        return sock

    def send_flags(self, flags: List[Flag]) -> List[Flag]:
        protocol_config = self.config.ctf.protocol

        if not flags:
            logging.debug("Нет флагов для отправки.")
            return []

        try:
            with self.connect_to_checksystem(
                protocol_config.checksys_host, protocol_config.checksys_port
            ) as sock:
                responses: List[str] = []
                try:
                    for flag in flags:
                        sock.sendall(flag.flag.encode("utf-8") + b"\n")
                        logging.info(f"Отправлен флаг: {flag.flag}")

                        if not (data := sock.recv(4096)):
                            break

                        messages = data.decode("utf-8").split("\n")
                        valid_messages = list(filter(None, messages))
                        responses.extend(valid_messages)

                        if len(responses) >= len(flags):
                            break

                except socket.timeout:
                    logging.info("Таймаут ожидания ответов от журейки.")

                # Обработка ответов и обновление статусов
                flags_to_update = flags[: len(responses)]
                for flag, response in zip(flags_to_update, responses):
                    if "Accepted" in response:
                        flag.status = FlagStatus.ACCEPTED
                    elif "Denied" in response:
                        flag.status = FlagStatus.REJECTED
                    flag.checksystem_response = response

                    if "Accepted" in response or "Denied" in response:
                        logging.info(f"Ответ от журейки: {response}")
                    else:
                        logging.warning(f"Неожиданный ответ: {response}")

                return flags_to_update

        except socket.error as e:
            logging.error(f"Ошибка сокета: {e}")
        except Exception as e:
            logging.error(f"Произошла ошибка при отправке флагов: {e}")

        return []
